maro/rl/policy/policy_manager.py

```diff
@@ -339,3 +339,193 @@ def exit(self):
+                quota = max(1, int(round(quota)))
+                policy_quota[name] = quota
+
+            # adjust quota if any redundancy occurs.
+            redundancy = self.num_trainers - sum(policy_quota.values())
+            if redundancy > 0:
+                busiest_policy = max(policy_quota, key=lambda name: policy_quota[name])
+                policy_quota[busiest_policy] += redundancy
+
+            for name, quota in policy_quota.items():
+                self._logger.info(
+                    f"policy {name} payload: {self._num_experiences_by_policy[name]},  quota: {quota} node(s)")
+                for i in range(quota):
+                    trainer_id = (i + offset) % self.num_trainers
+                    self._policy2trainer[name].append(f"TRAINER.{trainer_id}")
+                    self._trainer2policies[f"TRAINER.{trainer_id}"].append(name)
+                offset = (offset + quota) % self.num_trainers
+
+        # re-allocation
+        self._logger.info("Re-allocating policy states on trainers...")
+        for trainer_name, policy_names in self._trainer2policies.items():
+            self._proxy.send(
+                SessionMessage(
+                    MsgTag.INIT_POLICY_STATE, self._proxy.name, trainer_name,
+                    body={MsgKey.POLICY_STATE: {name: self.policy_dict[name].get_state() for name in policy_names}}
+                )
+            )
+
+    def update(self, exp_by_policy: Dict[str, ExperienceSet]):
+        exp_to_send, updated = {}, set()
+        for policy_name, exp in exp_by_policy.items():
+            self._num_experiences_by_policy[policy_name] += exp.size
+            self._exp_cache[policy_name].extend(exp)
+            if (
+                self._exp_cache[policy_name].size >= self.update_trigger[policy_name] and
+                self._num_experiences_by_policy[policy_name] >= self.warmup[policy_name]
+            ):
+                exp_to_send[policy_name] = self._exp_cache.pop(policy_name)
+                updated.add(policy_name)
+
+        # 1. prepare exp data for each trainer node
+        msg_body_by_dest = defaultdict(dict)
+        for policy_name, exp in exp_to_send.items():
+            trainer_id_list = self._policy2trainer[policy_name]
+            if len(trainer_id_list) == 1:  # single node
+                trainer_id = trainer_id_list[0]
+                if MsgKey.EXPERIENCES not in msg_body_by_dest[trainer_id]:
+                    msg_body_by_dest[trainer_id][MsgKey.EXPERIENCES] = {}
+                msg_body_by_dest[trainer_id][MsgKey.EXPERIENCES][policy_name] = exp
+                self._logger.info(f'policy {policy_name}, exp.size = {exp.size}')
+            else:
+                for i, trainer_id in enumerate(trainer_id_list):
+                    if MsgKey.EXPERIENCES not in msg_body_by_dest[trainer_id]:
+                        msg_body_by_dest[trainer_id][MsgKey.EXPERIENCES] = {}
+                    sub_exp = exp[i::len(trainer_id_list)]
```

Comment:
I was wondering if continuous slides would be faster?

Reply:
Maybe, but  slice with stride will be convenient and flexible when the `exp.size` is not divisible, I guess.